File: apps/mobile/lib/services/scenario_narrator_service.dart
// Scenario Narrator Service — Sprint S37.
//
// Narrates 3 retirement scenarios (prudent/base/optimiste) as
// educational text anchored on concrete numbers.
//
// Without BYOK: deterministic structured text.
// With BYOK (future): LLM generates -> ComplianceGuard validates.
//
// Each narrative: max 150 words.
// Each MUST mention return assumption and uncertainty.
// No prescriptive language, no banned terms.
// All French, informal "tu".

// ────────────────────────────────────────────────────────────
//  DATA CLASSES
// ────────────────────────────────────────────────────────────

/// A single narrated scenario with its key figures and text.
class NarratedScenario {
  final String label;

  /// Educational narrative text (max 150 words).
  final String narrative;
  final double annualReturnPct;
  final double capitalFinal;
  final double monthlyIncome;

  const NarratedScenario({
    required this.label,
    required this.narrative,
    required this.annualReturnPct,
    required this.capitalFinal,
    required this.monthlyIncome,
  });
}

/// Result containing 3 narrated scenarios + compliance fields.
class ScenarioNarrationResult {
  final List<NarratedScenario> scenarios;
  final String disclaimer;
  final List<String> sources;

  const ScenarioNarrationResult({
    required this.scenarios,
    required this.disclaimer,
    required this.sources,
  });
}

// ────────────────────────────────────────────────────────────
//  SERVICE
// ────────────────────────────────────────────────────────────

/// Deterministic scenario narrator.
///
/// Takes ForecasterService output figures and produces
/// educational narrative text for each of the 3 scenarios.
class ScenarioNarratorService {
  ScenarioNarratorService._();

  /// Narrate 3 scenarios from ForecasterService output.
  ///
  /// [prudentCapital] / [prudentMonthly] — prudent scenario figures.
  /// [baseCapital] / [baseMonthly] — base scenario figures.
  /// [optimisteCapital] / [optimisteMonthly] — optimiste scenario figures.
  /// [firstName] — user first name for personalisation.
  static ScenarioNarrationResult narrate({
    required double prudentCapital,
    required double prudentMonthly,
    required double baseCapital,
    required double baseMonthly,
    required double optimisteCapital,
    required double optimisteMonthly,
    String firstName = 'utilisateur',
  }) {
    final scenarios = <NarratedScenario>[
      NarratedScenario(
        label: 'Scenario prudent',
        annualReturnPct: 1.0,
        capitalFinal: prudentCapital,
        monthlyIncome: prudentMonthly,
        narrative: _narratePrudent(
          capital: prudentCapital,
          monthly: prudentMonthly,
          firstName: firstName,
        ),
      ),
      NarratedScenario(
        label: 'Scenario de reference',
        annualReturnPct: 4.5,
        capitalFinal: baseCapital,
        monthlyIncome: baseMonthly,
        narrative: _narrateBase(
          capital: baseCapital,
          monthly: baseMonthly,
          firstName: firstName,
        ),
      ),
      NarratedScenario(
        label: 'Scenario favorable',
        annualReturnPct: 7.0,
        capitalFinal: optimisteCapital,
        monthlyIncome: optimisteMonthly,
        narrative: _narrateOptimiste(
          capital: optimisteCapital,
          monthly: optimisteMonthly,
          firstName: firstName,
        ),
      ),
    ];

    return ScenarioNarrationResult(
      scenarios: scenarios,
      disclaimer:
          'Outil éducatif — ne constitue pas un conseil financier au sens '
          'de la LSFin. Les projections reposent sur des hypothèses de '
          'rendement et ne présagent pas des résultats futurs. '
          'Consulte un·e spécialiste pour un plan personnalisé.',
      sources: [
        'LAVS art. 21-29 (rente AVS)',
        'LPP art. 14 (taux de conversion)',
        'OPP3 art. 7 (plafond 3a)',
        'LPP art. 15-16 (bonifications)',
      ],
    );
  }

  // ════════════════════════════════════════════════════════════════
  //  NARRATIVE TEMPLATES
  // ════════════════════════════════════════════════════════════════

  static String _narratePrudent({
    required double capital,
    required double monthly,
    required String firstName,
  }) {
    final chfCapital = _formatChf(capital);
    final chfMonthly = _formatChf(monthly);
    return 'Scenario prudent (rendement ~1%/an) : '
        '$firstName, avec un capital estime a CHF $chfCapital a la retraite, '
        'ta rente mensuelle pourrait avoisiner CHF $chfMonthly. '
        'Ce scenario suppose une croissance modeste de tes avoirs, '
        'proche des taux obligatoires LPP actuels. '
        "L'incertitude reste presente — les marches peuvent evoluer "
        'differemment et les conditions economiques changer au fil des annees.';
  }

  static String _narrateBase({
    required double capital,
    required double monthly,
    required String firstName,
  }) {
    final chfCapital = _formatChf(capital);
    final chfMonthly = _formatChf(monthly);
    return 'Scenario de reference (rendement ~4.5%/an) : '
        '$firstName, le capital estime atteindrait CHF $chfCapital, '
        'soit environ CHF $chfMonthly/mois de revenu a la retraite. '
        'Ce scenario repose sur des hypotheses medianes historiques. '
        'Les resultats reels dependront de nombreux facteurs : '
        'inflation, performances des marches, evolution de ta carriere '
        'et de tes versements.';
  }

  static String _narrateOptimiste({
    required double capital,
    required double monthly,
    required String firstName,
  }) {
    final chfCapital = _formatChf(capital);
    final chfMonthly = _formatChf(monthly);
    return 'Scenario favorable (rendement ~7%/an) : '
        '$firstName, CHF $chfCapital en capital, '
        'soit environ CHF $chfMonthly/mois. '
        'Ce scenario suppose des conditions de marche favorables '
        'sur une longue periode. Les projections restent des estimations '
        '— aucun rendement futur ne peut etre predit avec exactitude. '
        'Diversification et discipline restent des leviers importants.';
  }

  // ════════════════════════════════════════════════════════════════
  //  FORMATTING HELPERS
  // ════════════════════════════════════════════════════════════════

  /// Format a number with Swiss apostrophe grouping (e.g. 1'250'000).
  static String _formatChf(double value) {
    final intVal = value.round();
    final str = intVal.abs().toString();
    final buffer = StringBuffer();
    if (intVal < 0) buffer.write('-');
    for (int i = 0; i < str.length; i++) {
      if (i > 0 && (str.length - i) % 3 == 0) {
        buffer.write("'");
      }
      buffer.write(str[i]);
    }
    return buffer.toString();
  }
}

File: services/backend/app/services/scenario/scenario_narrator_service.py
"""
Scenario Narrator Service — Sprint S37.

Narrates 3 retirement scenarios (prudent / base / optimiste) as educational text.

Without BYOK: structured deterministic text with numbers.
With BYOK: LLM narrates (future — ComplianceGuard validates).

Rules:
    - Each scenario: max 150 words
    - Each MUST mention the return assumption
    - Each MUST mention uncertainty
    - No prescriptive language
    - No banned terms (garanti, certain, assure, sans risque, optimal, meilleur,
      parfait, conseiller, tu devrais, tu dois, il faut que tu)
    - CHF formatted with Swiss apostrophe (e.g. 123'456)
    - All French, informal "tu"

Sources:
    - LPP art. 14-16 (prevoyance professionnelle)
    - LAVS art. 21-40 (AVS)
    - LSFin art. 3 (information financiere)
"""

from typing import List

from app.services.scenario.scenario_models import (
    NarratedScenario,
    ScenarioInput,
    ScenarioNarrationResult,
)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

STANDARD_DISCLAIMER = (
    "Outil educatif simplifie. Ne constitue pas un conseil financier (LSFin). "
    "Les projections reposent sur des hypotheses et ne representent pas "
    "des rendements futurs."
)

STANDARD_SOURCES = [
    "LSFin art. 3",
    "LPP art. 14-16",
    "LAVS art. 21-40",
]

# Banned terms — must NEVER appear in any narrative output
BANNED_TERMS = [
    "garanti",
    "certain",
    "assuré",
    "sans risque",
    "optimal",
    "meilleur",
    "parfait",
    "conseiller",
    "tu devrais",
    "tu dois",
    "il faut que tu",
]

# Narrative templates per scenario label
_TEMPLATES = {
    "prudent": (
        "Scenario prudent (rendement {pct}%/an) : avec un capital estime "
        "a CHF {capital} a la retraite, ta rente mensuelle pourrait avoisiner "
        "CHF {monthly}. Ce scenario suppose une croissance modeste. "
        "L'incertitude reste presente — les marches peuvent evoluer differemment."
    ),
    "base": (
        "Scenario de reference (rendement {pct}%/an) : le capital estime "
        "atteindrait CHF {capital}, soit environ CHF {monthly}/mois. "
        "Ce scenario repose sur des hypotheses medianes. "
        "Les resultats reels dependront de nombreux facteurs."
    ),
    "optimiste": (
        "Scenario favorable (rendement {pct}%/an) : CHF {capital} en capital, "
        "soit environ CHF {monthly}/mois. Ce scenario suppose des conditions "
        "de marche favorables. Aucun rendement n'est acquis au sens strict "
        "— les projections restent des estimations."
    ),
}

# Fallback for unknown labels
_FALLBACK_TEMPLATE = (
    "Scenario {label} (rendement {pct}%/an) : capital estime a "
    "CHF {capital}, revenu mensuel estime a CHF {monthly}. "
    "Les projections sont indicatives et soumises a l'incertitude des marches."
)


def _format_chf(amount: float) -> str:
    """Format a CHF amount with Swiss apostrophe separators.

    Examples:
        1234.56   -> "1'235"
        123456.78 -> "123'456"
        1000000   -> "1'000'000"
    """
    rounded = round(amount)
    if rounded < 0:
        return "-" + _format_chf(-rounded)
    s = str(rounded)
    # Insert apostrophes from the right every 3 digits
    parts = []
    while len(s) > 3:
        parts.append(s[-3:])
        s = s[:-3]
    parts.append(s)
    return "'".join(reversed(parts))


class ScenarioNarratorService:
    """Narrates 3 retirement scenarios as educational text.

    Deterministic fallback mode — no LLM required.
    """

    def narrate_scenarios(
        self,
        scenarios: List[ScenarioInput],
        first_name: str = "utilisateur",
        age: int = 30,
    ) -> ScenarioNarrationResult:
        """Generate narratives for the given scenarios.

        Args:
            scenarios: List of ScenarioInput (typically 3: prudent/base/optimiste).
            first_name: User's first name for personalisation (future LLM use).
            age: User's current age (future LLM use).

        Returns:
            ScenarioNarrationResult with narrated scenarios, disclaimer, sources.
        """
        narrated: List[NarratedScenario] = []

        for sc in scenarios:
            pct = round(sc.annual_return * 100, 1)
            capital_str = _format_chf(sc.capital_final)
            monthly_str = _format_chf(sc.monthly_income)

            template = _TEMPLATES.get(sc.label, _FALLBACK_TEMPLATE)
            narrative = template.format(
                pct=pct,
                capital=capital_str,
                monthly=monthly_str,
                label=sc.label,
            )

            # Safety: strip any accidental banned terms
            narrative = self._sanitize(narrative)

            narrated.append(
                NarratedScenario(
                    label=sc.label,
                    narrative=narrative,
                    annual_return_pct=pct,
                    capital_final=sc.capital_final,
                    monthly_income=sc.monthly_income,
                )
            )

        return ScenarioNarrationResult(
            scenarios=narrated,
            disclaimer=STANDARD_DISCLAIMER,
            sources=list(STANDARD_SOURCES),
            uncertainty_mentioned=True,
        )

    @staticmethod
    def _sanitize(text: str) -> str:
        """Remove banned terms from generated text.

        This is a safety net — templates should already be clean.
        Replaces banned terms with safe alternatives.
        """
        replacements = {
            "garanti au sens strict": "acquis",
            "garanti": "acquis",
            "certain": "probable",
            "assuré": "estime",
            "sans risque": "a faible risque",
            "optimal": "adapte",
            "meilleur": "favorable",
            "parfait": "adequat",
            "conseiller": "specialiste",
            "tu devrais": "tu pourrais",
            "tu dois": "tu peux",
            "il faut que tu": "tu peux envisager de",
        }
        lower = text.lower()
        for banned, safe in replacements.items():
            if banned in lower:
                # Case-insensitive replacement
                import re

                text = re.sub(re.escape(banned), safe, text, flags=re.IGNORECASE)
        return text
